File: 501-FindModeInBST/inOrder.cpp
#include <iostream>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
private:
    vector<int> res;
    int pre;
    int cur;
    int curSum;
    int maxSum;
public:
    vector<int> findMode(TreeNode* root) {
        maxSum = 1;
        traverse(root);
        return res;
    }

    void traverse(TreeNode* root) {
        if (root == nullptr) return;
        traverse(root->left);
        cur = root->val;
        if (curSum && cur == pre) curSum++;
        else {
            pre = cur;
            curSum = 1;
        }

        if (curSum > maxSum) {
            res.clear();
            maxSum = curSum;
            res.push_back(cur);
        }
        else if (curSum == maxSum) {
            res.push_back(cur);
        }
        traverse(root->right);
    }
};

File: 98-ValidateBST/inOrder.java
class Solution {
    private long lastVal = Long.MIN_VALUE;
    //比较过程:
    //lastVal = root的左子树中最右子树的值
    //比较lastVal与root.val
    //lastVal = root.val
    //比较lastVal与root右子树中最左子树的值

    public boolean isValidBST(TreeNode root) {
        if (root == null)
            return true;
        if (!isValidBST(root.left))
            return false;
        if (root.val <= lastVal) 
            return false;
        lastVal = (long) root.val;
        if (!isValidBST(root.right))
            return false;
        return true;
    }
}